### DISCUSSION TITLE
Next.js on Netlify Edge Functions Beta discussion

### DISCUSSION THREAD

**Discussion by ascorbic** [opennextjs/opennextjs-netlify#1319] [Q&A]
We now have support for deploying Next.js sites to the Netlify Edge Functions Beta. 
It includes support for deploying middleware to the edge, but also for deploying your site to the edge so you can use streaming with React Server Components. See [the React Server Components demo](https://github.com/netlify/next-react-server-components) for more info.

Please share your feedback here!

**Comment by JohnGemstone** [comment]
Huge step! Thanks for everyone's work on this

**Reply by adam-cohen-amplitude** [reply]
When will this be out of Beta?

**Reply by charkour** [reply]
It's currently out of beta and is now the default runtime! Hope this helps.

**Comment by Joroze** [comment]
What is the difference between using Netlify edge functions and built-in NextJS API routes?

https://nextjs.org/docs/api-routes/edge-api-routes

**Reply by charkour** [reply]
Hey @Joroze, the page you linked is documentation on how to use edge api routes with Next.js and Vercel Edge which is a slightly different API than using Next.js with [Netlify Edge](https://www.netlify.com/blog/next.js-middleware-on-netlify/).

**Reply by Joroze** [reply]
Makes sense. Thanks!

**Reply by charkour** [reply]
Haha of course. I was initially confused by this as well. It's not clear.

**Reply by ascorbic** [reply]
Yes. You _can_ use Next.js API routes with edge runtime though: we automatically include the compatibility layer, so you can use the Next.js syntax instead of the Netlify Edge Funciton syntax.

**Reply by weidezhang** [reply]
Hi ascorbic, 

I am using nextjs 13 edge api route. Is the process.env still valid if converted to netilfy edge runtime ?  I tried to use Netlify.env but seems not provided for NextJS API route. Does that mean I have to convert the api route to netilfy edge functions first ? 

And using netlify-cli, is there a way I can see what api route is converted to netilify edge functions what are converted to regular functions ? Thank you ,

**Comment by Jennievon** [comment]
Hi, I have the env variable - `NEXT_USE_NETLIFY_EDGE` - set to `true` for all contexts but my middleware seems not to be running. Any thoughts on why this could be?

**Reply by adam-cohen-amplitude** [reply]
If you're using matchers in the context object, the middleware won't run. Netlify doesn't support matchers, and middleware functions won't run if there is one in the file.

**Reply by ascorbic** [reply]
We do support matchers now, but I know there have been some issues with specific regex matchers. What matcher syntax are you using?

**Reply by Auspicus** [reply]
@ascorbic I found that `matcher: "/"` works fine locally but doesn't get picked up in Netlify. Could be user error but was not what I was expecting. Trying to handle this inside the middleware instead but running into a similar problem. Doesn't look like it's running at all. ie.

```js
export function middleware(request: NextRequest) {
  const parsedUrl = new URL(request.url)

  if (parsedUrl.pathname === '/') {
    // ...
  }

  return NextResponse.next()
}
```

and

```js
export function middleware(request: NextRequest) {
  // ...
}

export const config = {
  matcher: '/',
}
```

both don't work.

**Reply by Auspicus** [reply]
Figured it out. Turns out using "Access Control" like setting a password in Netlify alongside Edge Middleware doesn't work. I was able to disable the password protection I had on the site for a few minutes and that was all that was needed to get this working.

**Reply by Jennievon** [reply]
@Auspicus wondering how this works 'cause I do not have that password enabled on my end but I'm still unable to get it working.

**Reply by ascorbic** [reply]
Could anyone having trouble confirm if they have i18n enabled, and if so whether it works for non-default locales? I think I may have the answer.

**Reply by Auspicus** [reply]
@Jennievon I would go with the simplest config possible with a single middleware and work from there. I haven't tested the matcher config version since I disabled access-control and it has been working. Perhaps there's another step missing. Can you reproduce your issue with a minimal environment as an example?

**Reply by Auspicus** [reply]
@ascorbic We don't have any i18n enabled. I am longer having issues with the middleware after disabling the access control. I haven't tested with matcher config though.

**Reply by tony-cocco** [reply]
We are having issues with our build not respecting our matcher config with regex. i18n is enabled. Access Control is disabled. I am testing some other regex patterns; when I tried '(/.*)' everything was timing out, so I assume it was (kinda) working. Our original implementation was using a negative lookahead as described in the nextjs docs, currently trying with some other patterns but not having any luck.

**Reply by ascorbic** [reply]
Hi @tony-cocco and everyone
I've been [working on a fix](https://github.com/netlify/next-runtime/pull/1693) and should have it ready to ship today. Thanks for your patience.

**Reply by ascorbic** [reply]
This fix has now rolled out in version 4.27.3 of the runtime. If you do a new build it should now work. Please report back to let me know if it works or not!

**Reply by tony-cocco** [reply]
@ascorbic I readded our middleware file and redeployed. I did confirm the runtime from the build logs was 4.27.3. After the deploy, the netlify handler logs were just spewing out task timeouts (exceeding 10s) before I really hit the application. Opening the app showed this error as well. I revised our regex to just be the example regex from the next docs (`'/((?!api|static|favicon.ico).*)'`). This didn't change any behavior. Keep in mind both of these work as expected when I run the server locally.

I noted during the build that netlify does a lot of route renaming from the looks of it. Do we need a better regex? Right now I assume we're infinitely looping given the timeouts.

Additionally, I don't know when the change happened but I noticed the netlify build is exploding all the api files as separate functions (I think this was the point of this beta though). This sends our build time average from ~2.5min to 5+ just for the function bundle step (none of those files were even changing). Not sure if that was intended or a separate discussion altogether.

Let me know if I can be of more assistance.

**Reply by ascorbic** [reply]
Thanks. Just to confirm, are the timeouts coming from the edge functions or regular Netlify functions logs?

**Reply by tony-cocco** [reply]
I was seeing them from hitting the deploy as well as `.netlify/functions/___netlify-handler` logs.

**Reply by ascorbic** [reply]
Does the same happen without the matcher?

**Reply by tony-cocco** [reply]
@ascorbic Deployed without the matcher config, same behavior.

**Reply by ascorbic** [reply]
Ok, what about without the middleware at all. Thanks for your help in diagnosing this

**Reply by tony-cocco** [reply]
Without the middleware is how we deploy currently, but we can't do what we want in the middleware without putting it on every page.

Our current goal was to redirect away from any of our "under development" pages. The middleware function inspects the destination pathname, and if it's a "hidden" path we redirect to /. Without the middleware, we forked our codebase and just deleted those files. In the future, we want to do some audit logging on routes which _should_ live in middleware, so getting this to work is of value to us. 

Happy to continue to assist. We love the platform so your success is shared. =)

**Reply by tony-cocco** [reply]
@ascorbic I want to add that explicit matchers works for us (to confirm one of the earlier commenters).

If I set matcher config to `['/thispath', '/orthispath']` navigating to those does trigger our middleware. Something about the loose-ness of the example negative lookahead just doesn't jive with the netlify environment.

**Reply by adam-cohen-amplitude** [reply]
I have not been able to get any matcher to work on Netlify. I was told by their support team they don't support lookahead functionality in middleware or edge functions.

**Reply by ascorbic** [reply]
Thanks. I'm mainly trying to confirm if right now you can deploy without middleware and then load the pages wihtout errors. I'm interested in whether something has broken in the recent update.

**Reply by tony-cocco** [reply]
Summary: All of our deployments are working as long as they don't use regex in middleware matchers. @ascorbic

**Reply by nickytonline** [reply]
I've created a minimal reproduction and can confirm it doesn't work when regex are used in middleware matchers. I'm currently looking into this.

**Reply by nickytonline** [reply]
Actually, it is working. I somehow had installed an older version of the plugin and didn't clue in. I'm currently on the version 4.27.3 of the plugin. Here's what my middleware looks like.

```js
import { MiddlewareRequest } from '@netlify/next';

export async function middleware(req) {
	const { pathname } = req.nextUrl;
	const request = new MiddlewareRequest(req);
	const res = await request.next();

	res.headers.set('x-show-pathname', pathname);
	res.headers.set('x-is-show', true);
	return res;
}

export const config = {
	matcher: [{ source: '/shows/((?!99|88).*)' }]
};

```

As long as the show isn't 88 or 99, you should see the `x-show` and `x-is-show` headers.
- https://6351b60978bc3f170e1d03c2--nextjs-middleware-matchers.netlify.app/shows/99 (no custom headers)
- https://6351b60978bc3f170e1d03c2--nextjs-middleware-matchers.netlify.app/shows/88 (no custom headers)
- https://6351b60978bc3f170e1d03c2--nextjs-middleware-matchers.netlify.app/shows/12 (yay! custom headers)

Here's the repository, https://github.com/nickytonline/matchers-with-regex-middleware-nextjs

![CleanShot 2022-10-20 at 17 05 14](https://user-images.githubusercontent.com/833231/197058017-5b997044-f0e3-4b0c-b75d-8aaa0439b120.png)

![CleanShot 2022-10-20 at 17 06 35](https://user-images.githubusercontent.com/833231/197058179-1ff07b27-0b43-4005-8bcf-bc2698bb2cab.png)

**Reply by tony-cocco** [reply]
@nickytonline Thanks for the working example! Can I ask you to try moving the lookahead up to the front? Our usage (and the one from the nextjs docs) begins with the `(?!)` pattern and my assumption is it's too aggressive and causing issues. So that any route that isn't `/show` triggers the middleware? The only other major difference I see is you're using netlify's request object. So I can try using that instead of the nextjs one in the meantime.

**Reply by tony-cocco** [reply]
Alrighty... this took so much time to debug. So here's some code to look at...

```typescript
import { NextRequest, NextResponse } from 'next/server';

export function middleware(request: NextRequest) {
  const url = request.nextUrl.clone();
  console.log( url.pathname );

  if (
    // ignore home page and any of the special urls, just continue
    url.pathname === '/' || url.pathname.match(/^\/api|assets|static|_next|.netlify|favicon.ico.*/)
  )
    return NextResponse.next();

  // redirect to "home" for anything else...
  url.pathname = '/';
  return NextResponse.redirect(url);
}

// export const config = {
//   matcher: [
//     // example from nextjs docs...
//     // '/((?!api|static|favicon.ico).*)', // <- works locally but not in netlify.
//   ],
// };
```

After playing around in regex101 for a while, I'm convinced whatever flavor of regex matcher uses is something else. It doesn't behave like I'd expect. So... I removed the matcher config and wrote the regex in the function... This works as I expect it to on local and in netlify.

In netlify, I noticed when I deployed the matcher with the example regex that `/api/auth/session` was triggering the middleware, even though, based on the commenting, urls starting with `/api` should be skipped. This is a special url from next-auth though... perhaps there is a conflicting or superseding rule for next-auth integration?

@nickytonline @ascorbic Thoughts welcome.
